File: artifacts/api-server/src/lib/delivery.ts
import { eq } from "drizzle-orm";
import { db, handoffs, conversations, settings, type Handoff } from "./db";
import { broadcastChat } from "./ws";
import { logger } from "./logger";

type DeliveryResult = { ok: boolean; detail?: string };

// Dispatch an agent reply to the originating channel. Web uses the chat WS;
// Telegram/WhatsApp call the respective HTTP APIs when credentials are set.
async function dispatchOnce(
  handoff: Handoff,
  sessionId: string,
  content: string,
): Promise<DeliveryResult> {
  if (handoff.channel === "web") {
    const ok = broadcastChat(sessionId, {
      type: "agent.message",
      handoffId: handoff.id,
      content,
      time: new Date().toISOString(),
    });
    // Web is "fire and forget"; treat as delivered even with no live client
    // because the widget will pick it up via /api/chat/:sessionId/agent-messages polling.
    return { ok: true, detail: ok ? "delivered live" : "queued for poll" };
  }

  // Look up per-tenant settings for channel creds.
  const [tenantSettings] = await db
    .select()
    .from(settings)
    .where(eq(settings.tenantId, handoff.tenantId))
    .limit(1);

  if (handoff.channel === "telegram") {
    const token = tenantSettings?.telegramToken;
    if (!token || !tenantSettings?.isTelegramActive) {
      return { ok: false, detail: "Telegram not configured" };
    }
    try {
      const res = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ chat_id: sessionId, text: content }),
      });
      if (!res.ok) return { ok: false, detail: `Telegram ${res.status}` };
      return { ok: true };
    } catch (err) {
      return { ok: false, detail: (err as Error).message };
    }
  }

  if (handoff.channel === "whatsapp") {
    const sid = tenantSettings?.twilioSid;
    const tok = tenantSettings?.twilioToken;
    const from = tenantSettings?.twilioNumber;
    if (!sid || !tok || !from || !tenantSettings?.isWhatsappActive) {
      return { ok: false, detail: "WhatsApp not configured" };
    }
    try {
      const body = new URLSearchParams({
        To: `whatsapp:${sessionId}`,
        From: `whatsapp:${from}`,
        Body: content,
      });
      const res = await fetch(
        `https://api.twilio.com/2010-04-01/Accounts/${sid}/Messages.json`,
        {
          method: "POST",
          headers: {
            Authorization: `Basic ${Buffer.from(`${sid}:${tok}`).toString("base64")}`,
            "Content-Type": "application/x-www-form-urlencoded",
          },
          body: body.toString(),
        },
      );
      if (!res.ok) return { ok: false, detail: `Twilio ${res.status}` };
      return { ok: true };
    } catch (err) {
      return { ok: false, detail: (err as Error).message };
    }
  }

  return { ok: false, detail: `Unknown channel: ${handoff.channel}` };
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

// 3-attempt retry with exponential backoff (1s/2s/4s). Updates delivery_status
// + delivery_attempts on every attempt so the dashboard sees real state.
export async function deliverHandoffReply(
  handoffId: number,
  content: string,
): Promise<DeliveryResult> {
  const [handoff] = await db.select().from(handoffs).where(eq(handoffs.id, handoffId)).limit(1);
  if (!handoff) return { ok: false, detail: "Handoff not found" };
  const [conv] = await db
    .select()
    .from(conversations)
    .where(eq(conversations.id, handoff.conversationId))
    .limit(1);
  const sessionId = conv?.customerRef || "";

  let last: DeliveryResult = { ok: false, detail: "no attempts" };
  for (let attempt = 1; attempt <= 3; attempt++) {
    last = await dispatchOnce(handoff, sessionId, content);
    await db
      .update(handoffs)
      .set({
        deliveryAttempts: attempt,
        deliveryStatus: last.ok ? "delivered" : "failed",
        deliveryDetail: last.detail ?? null,
      })
      .where(eq(handoffs.id, handoffId));
    if (last.ok) return last;
    if (attempt < 3) await sleep(1000 * 2 ** (attempt - 1));
  }
  logger.warn({ handoffId, last }, "Handoff delivery exhausted retries");
  return last;
}

File: artifacts/api-server/app/services/delivery.py
import asyncio
import base64
import json
import httpx
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.models import Handoff, Conversation, Settings
from app.websocket_manager import ws_manager


async def deliver_handoff_reply(db: AsyncSession, handoff_id: int, content: str) -> dict:
    result = await db.execute(select(Handoff).where(Handoff.id == handoff_id))
    handoff = result.scalar_one_or_none()
    if not handoff:
        return {"ok": False, "status": "error", "detail": "Handoff not found"}

    result = await db.execute(select(Conversation).where(Conversation.id == handoff.conversation_id))
    conv = result.scalar_one_or_none()
    if not conv:
        handoff.delivery_status = "failed"
        handoff.delivery_detail = "Conversation missing"
        await db.commit()
        return {"ok": False, "status": "error", "detail": "Conversation missing"}

    result = await db.execute(select(Settings).where(Settings.tenant_id == handoff.tenant_id))
    s = result.scalar_one_or_none()

    customer_ref = conv.customer_ref
    delays = [0, 1, 2]

    last_result: dict = {"ok": False, "status": "error", "detail": "No attempt made"}
    for attempt, delay in enumerate(delays):
        if delay:
            await asyncio.sleep(delay)
        handoff.delivery_attempts = attempt + 1
        try:
            last_result = await _dispatch_once(handoff, conv, s, customer_ref, content)
        except Exception as e:
            last_result = {"ok": False, "status": "error", "detail": str(e)}

        handoff.delivery_status = "delivered" if last_result["ok"] else "failed"
        handoff.delivery_detail = last_result.get("detail")
        await db.commit()

        if last_result["ok"]:
            break

    return last_result


async def _dispatch_once(handoff: Handoff, conv: Conversation | None, s: Settings | None, customer_ref: str, content: str) -> dict:
    channel = handoff.channel

    if channel == "web":
        delivered = await ws_manager.broadcast_chat(customer_ref, {"type": "agent_message", "content": content})
        return {"ok": delivered, "status": "delivered" if delivered else "pending", "detail": None}

    if channel == "telegram":
        if not s or not s.telegram_token or not s.is_telegram_active:
            return {"ok": False, "status": "error", "detail": "Telegram not configured"}
        url = f"https://api.telegram.org/bot{s.telegram_token}/sendMessage"
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(url, json={"chat_id": customer_ref, "text": content})
            ok = resp.status_code == 200
            return {"ok": ok, "status": "delivered" if ok else "failed", "detail": resp.text if not ok else None}

    if channel == "whatsapp":
        if not s or not s.twilio_sid or not s.twilio_token or not s.twilio_number or not s.is_whatsapp_active:
            return {"ok": False, "status": "error", "detail": "WhatsApp not configured"}
        url = f"https://api.twilio.com/2010-04-01/Accounts/{s.twilio_sid}/Messages.json"
        creds = base64.b64encode(f"{s.twilio_sid}:{s.twilio_token}".encode()).decode()
        data = {"To": f"whatsapp:{customer_ref}", "From": f"whatsapp:{s.twilio_number}", "Body": content}
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(url, data=data, headers={"Authorization": f"Basic {creds}"})
            ok = resp.status_code in (200, 201)
            return {"ok": ok, "status": "delivered" if ok else "failed", "detail": resp.text if not ok else None}

    return {"ok": False, "status": "error", "detail": f"Unknown channel: {channel}"}
